src/language.rs

```diff
@@ -11,115 +12,65 @@ pub enum SupportedLanguageName {
-
-impl SupportedLanguage for SupportedLanguageRust {
-    fn language(&self) -> Language {
-        tree_sitter_rust::language()
-    }
-
-    fn name(&self) -> SupportedLanguageName {
-        SupportedLanguageName::Rust
-    }
-
-    fn name_for_ignore_select(&self) -> &'static str {
-        "rust"
-    }
-
-    fn extensions(&self) -> Vec<&'static str> {
-        vec!["rs"]
-    }
-}
-
-pub fn get_rust_language() -> SupportedLanguageRust {
-    SupportedLanguageRust
-}
-
-pub struct SupportedLanguageTypescript;
-
-impl SupportedLanguage for SupportedLanguageTypescript {
-    fn language(&self) -> Language {
-        tree_sitter_typescript::language_tsx()
-    }
-
-    fn name(&self) -> SupportedLanguageName {
-        SupportedLanguageName::Typescript
-    }
-
-    fn name_for_ignore_select(&self) -> &'static str {
-        "ts"
-    }
-
-    fn extensions(&self) -> Vec<&'static str> {
-        vec!["ts", "tsx"]
-    }
-}
-
-pub fn get_typescript_language() -> SupportedLanguageTypescript {
-    SupportedLanguageTypescript
-}
-
-pub struct SupportedLanguageJavascript;
-
-impl SupportedLanguage for SupportedLanguageJavascript {
-    fn language(&self) -> Language {
-        tree_sitter_javascript::language()
-    }
-
-    fn name(&self) -> SupportedLanguageName {
-        SupportedLanguageName::Javascript
-    }
-
-    fn name_for_ignore_select(&self) -> &'static str {
-        "js"
-    }
-
-    fn extensions(&self) -> Vec<&'static str> {
-        vec!["js", "jsx", "vue", "cjs", "mjs"]
+impl PartialEq for SupportedLanguage {
+    fn eq(&self, other: &Self) -> bool {
+        self.language == other.language
     }
 }
 
-pub fn get_javascript_language() -> SupportedLanguageJavascript {
-    SupportedLanguageJavascript
-}
-
-pub fn get_all_supported_languages() -> HashMap<SupportedLanguageName, Box<dyn SupportedLanguage>> {
-    [
-        (
-            SupportedLanguageName::Rust,
-            Box::new(get_rust_language()) as Box<dyn SupportedLanguage>,
-        ),
-        (
-            SupportedLanguageName::Typescript,
-            Box::new(get_typescript_language()) as Box<dyn SupportedLanguage>,
-        ),
-        (
-            SupportedLanguageName::Javascript,
-            Box::new(get_javascript_language()) as Box<dyn SupportedLanguage>,
-        ),
+pub static ALL_SUPPORTED_LANGUAGES: Lazy<Vec<SupportedLanguage>> = Lazy::new(|| {
+    vec![
+        SupportedLanguage {
+            language: tree_sitter_rust::language(),
+            name: SupportedLanguageName::Rust,
+            name_for_ignore_select: "rust",
+        },
+        SupportedLanguage {
+            language: tree_sitter_typescript::language_tsx(),
+            name: SupportedLanguageName::Typescript,
+            name_for_ignore_select: "ts",
+        },
+        SupportedLanguage {
+            language: tree_sitter_javascript::language(),
+            name: SupportedLanguageName::Javascript,
+            name_for_ignore_select: "js",
+        },
```

Comment:
This seems pretty macro-able if you want to eventually add all the languages and not type this out in such a verbose way.

Reply:
Or what about just a typical `SupportedLanguage::new()` constructor?

Or were you picturing stripping it down even more than that eg omitting the `SupportedLanguageName::` prefix (which I guess you could achieve via a `use SupportedLanguageName::*;` but could also achieve via an un-hygienic macro (?))?

Reply:
Yeah, that's what I was thinking, assuming you always use the same names as tree sitter does internally, but that's probably not worth it over just using a constructor.